added signupLanguage query support for registration

pnoll1 opened the pull request:
added support for signupLanguage for register and register with code routes.  register with code route has not been tested.

Sets locale cookie if signupLanguage query is present and is a valid language, otherwise does nothing.

All tests passing.

eloquence commented on routes/actions.js:
```diff
@@ -210,7 +219,16 @@ router.get('/register', function(req, res) {
+        });
+    i18n.setLocale(req, lang);
+  }
   InviteLink
```
Per DRY (don't repeat yourself) principle, I would suggest removing this into its own dedicated function so we don't have an exact copy of the same code in two places in the same file.

eloquence commented on routes/actions.js:
```diff
@@ -201,6 +201,15 @@ router.get('/new/user', function(req, res) {
 });
 
 router.get('/register', function(req, res) {
+  if (req.query.signupLanguage && languages.isValid(req.query.signupLanguage)){
```
We should add `&& !req.user` to this check so a malicious user can't switch someone else's language by sending them to a signupLanguage register link. This will simply suppress the feature if the user is already logged in.

pnoll1 commented:
corrected per comments.  put code in function and added check for signed in user.

manual testing shows code still works as intended and user check works.

all tests passing.

eloquence commented:
Thanks, @pnoll1, this looks pretty solid!

Thinking about it a bit more, the risk of having a malicious page contain a `<script>` tag which in turn loads a resource like `http://localhost/register?returnTo=/&signupLanguage=de` is still too high. While we're protecting logged in users now, this would set the language for a logged out user who is unaware of this change. 

How do we avoid this? I would suggest that we don't set the cookie until the form is submitted. If the user submits, regardless of whether successful or not, we can reasonably assume that they're OK with the language change. So in the POST handler, we could set the cookie - we'd just have to pass it along in the form, as we do with `returnTo`. 

Does that make sense? I can also do that part, if you prefer.

eloquence commented:
Yup, I think we can even safely get rid of the req.user check if the language change is initially not persistent. It may be useful for logged in folks to check that the link they're giving to someone else is producing the expected result.

pnoll1 commented:
Language is now set directly in render route if signupLang present and passes variables to POST registration request using same method as returnTo.  Cookie is set if user completes POST.

Removed user sign in check,  if we decide to add later, it should go in render to ensure signup query has no impact.

All tests passing.

eloquence commented on routes/actions.js:
```diff
@@ -210,7 +213,7 @@ router.get('/register', function(req, res) {
 
 router.get('/register/:code', function(req, res, next) {
-  const { code } = req.params;
+  const { code } = req.params.code;
```
I'm assuming this is an accidental/unintentional change. It breaks the invite link logic: the original assignment is a destructuring assignment; your change will always result in `undefined`. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring

pnoll1 replied:
semi-unintentional,  you tend to mix params and query when talking about scope so I changed this early on thinking I would need to differentiate params.

eloquence commented on routes/helpers/render.js:
```diff
@@ -14,6 +15,11 @@ let render = {
+    // override locale cookie using signupLanguage if present
+    if (req.query.signupLanguage)
+        i18n.setLocale(req, req.query.signupLanguage);
+        vars.signupLanguage = req.query.signupLanguage;
```
Nit: You'll either want to surround the entire block in `{` and `}` braces, or unindent the `vars.signupLanguage` statement. The `if` statement ends with the first semicolon, so the continued indentation is semantically incorrect (if meaningless). I would suggest surrounding the entire block with braces, though it functionally makes no difference whether the `vars.signupLanguage` statement always or sometimes executes.

eloquence commented on routes/actions.js:
```diff
@@ -361,4 +366,19 @@ function returnToPath(req, res) {
+// check for signupLanguage, ensure valid lang
+// set locale cookie if conditions met, else do nothing
+function setSignupLanguage(req, res) {
+    if (req.body.signupLanguage && languages.isValid(req.body.signupLanguage)) {
```
Nit: `if` block should be indented. (I tend to use a beautifier to do this kind of clean-up automatically within the editor; I recommend http://atom.io/ . It also lets you use ESLint, which will flag violations of the style rules defined for this repository.)

pnoll1 commented:
Fixed per comments.

Should probably add a test for registration with code.  I don't have a way to test that on my dev site and it's currently how everyone registers.
